src/utils/IdealGraphVisualizer/HierarchicalLayout/src/main/java/com/sun/hotspot/igv/hierarchicallayout/FreeInteractiveLayoutManager.java

```diff
@@ -0,0 +1,324 @@
+                if (layoutNodes.containsKey(neighborVertex)) {
+                    assignedNeighbors.add(layoutNodes.get(neighborVertex));
+                }
+            }
+
+            if (!assignedNeighbors.isEmpty()) {
+                if (assignedNeighbors.size() == 1) {
+                    // Single neighbor: position around the neighbor
+                    setPositionAroundSingleNode(node, assignedNeighbors.get(0), DISPLACEMENT_RANGE_SINGLE);
+                } else {
+                    // Multiple neighbors: Calculate barycenter with displacement
+                    calculateBarycenterWithDisplacement(node, assignedNeighbors, DISPLACEMENT_RANGE_BARYCENTER);
+                }
+            } else {
+                // No neighbors: Position randomly around (0, 0)
+                setRandomPositionAroundOrigin(node, random);
+            }
+
+            // Add the new node to the layout
+            layoutNodes.put(vertex, node);
+        }
+
+        // Second pass: Refine positions based on neighbor degree
+        newLayoutNodes.sort(LAYOUT_NODE_DEGREE_COMPARATOR.reversed());
+
+        // Collect all nodes (existing and new)
+        Collection<LayoutNode> allNodes = layoutNodes.values();
+
+        for (LayoutNode node : newLayoutNodes) {
+            Vertex vertex = node.getVertex();
+
+            // Gather assigned neighbors
+            List<LayoutNode> assignedNeighbors = new ArrayList<>();
+            for (Vertex neighborVertex : prevGraph.getNeighborVertices(vertex)) {
+                if (layoutNodes.containsKey(neighborVertex)) {
+                    assignedNeighbors.add(layoutNodes.get(neighborVertex));
+                }
+            }
+
+            if (!assignedNeighbors.isEmpty()) {
+                // Refine position based on force-based method
+                applyForceBasedAdjustment(node, assignedNeighbors, allNodes);
+            }
+
+            // Ensure node's position remains updated in the layout
+            layoutNodes.put(vertex, node);
+        }
+    }
+
+    private void applyForceBasedAdjustment(LayoutNode node, List<LayoutNode> assignedNeighbors, Collection<LayoutNode> allNodes) {
+        // Constants for force-based adjustment
+        final int ITERATIONS = 50;
+        final double REPULSION_CONSTANT = 1000;
+        final double SPRING_CONSTANT = 0.2;
+        final double DAMPING = 0.8;
+        final double IDEAL_LENGTH = 100;
+
+        double posX = node.getX();
+        double posY = node.getY();
+        double dx = 0, dy = 0; // Displacement
+
+        for (int i = 0; i < ITERATIONS; i++) {
+            double netForceX = 0;
+            double netForceY = 0;
+
+            // Repulsive forces from all other nodes
+            for (LayoutNode otherNode : allNodes) {
+                if (otherNode == node) continue; // Skip self
+
+                double deltaX = posX - otherNode.getX();
+                double deltaY = posY - otherNode.getY();
```

Comment:
What happens if this distance is zero? Does the division below behave ok?

Reply:
If we get issues here, we can always check for zero and add some random non-zero noise to force different position to get the two nodes to separate in a sane way.

Reply:
right! It should be fixed now